File: services/typescript/src/handler/error-mapping.ts
import { ConnectError, Code } from "@connectrpc/connect";
import {
  NotFoundError,
  InvalidInputError,
  AlreadyExistsError,
  ConflictError,
} from "../repository/errors.js";

export function toConnectError(err: unknown): ConnectError {
  if (err instanceof NotFoundError) {
    return new ConnectError(err.message, Code.NotFound);
  }
  if (err instanceof InvalidInputError) {
    return new ConnectError(err.message, Code.InvalidArgument);
  }
  if (err instanceof AlreadyExistsError || err instanceof ConflictError) {
    return new ConnectError(err.message, Code.AlreadyExists);
  }
  if (err instanceof Error) {
    return new ConnectError(err.message, Code.Internal);
  }
  return new ConnectError(String(err), Code.Internal);
}

File: services/kotlin/src/main/kotlin/com/corestack/handler/ErrorMapping.kt
package com.corestack.handler

import com.corestack.connect.ConnectCode
import com.corestack.connect.ConnectError
import com.corestack.repository.*

fun toConnectError(e: Exception): ConnectError = when (e) {
    is NotFoundError -> ConnectError(ConnectCode.NOT_FOUND, e.message ?: "not found")
    is InvalidInputError -> ConnectError(ConnectCode.INVALID_ARGUMENT, e.message ?: "invalid input")
    is AlreadyExistsError -> ConnectError(ConnectCode.ALREADY_EXISTS, e.message ?: "already exists")
    is ConflictError -> ConnectError(ConnectCode.ALREADY_EXISTS, e.message ?: "conflict")
    is ConnectError -> e
    else -> ConnectError(ConnectCode.INTERNAL, e.message ?: "internal error")
}
